### ISSUE TITLE
Add an IsRequired field to the EntityField Entity

### ISSUE THREAD

**Issue by JS-BC** [MemberJunction/MJ#593]
**Is your feature request related to a problem? Please describe.**
I want a way to prevent users from saving entity objects with certain fields set as an empty string, specifically fields that are an entity's name field, i.e. where `IsNameField` is set to 1

**Describe the solution you'd like**
- Add an `IsRequired` column to the `EntityField` entity
- During the validation step before saving an entity, if a field is marked as `IsRequired` then there will be logic to ensure the given field is not empty
- If an entity field is marked as required, then the `AllowsNull` field must enabled as well

**Describe alternatives you've considered**
- A user could subclass an entity and overwrite the Validate method to ensure specific fields are not empty

**Additional context**

**Comment by AN-BC** [comment]
@JS-BC  why is this needed in addition to the `AllowNull` attribute that is auto-derived from the nullability of the underlying column?

**Comment by JS-BC** [comment]
@AN-BC Yes. Specifically, I think fields marked as Name fields should not be allowed to be null or empty

**Comment by AN-BC** [comment]
I get that but perhaps there's a better way to handle this instead of a seemingly overlapping attribute of `IsRequired` what if we introduced 4 new attributes of the `EntityField` table:

- `MinLength` applies to text type fields only and specifies the minimum length of a value - could be 1 or a higher number and knock out this issue but also other similar things where you might want at least 3 or whatever characters
- `MaxLength` similar to Min but specifies the upper bound on field length
- `MinValue` for Date and Numeric type values, specifies the minimum value allowed for a field
- `MaxValue` same as MinValue but upper bound (optional)


This implementation would solve for your needs without a duplicative(sounding) new attribute and provide additional functionality, be easy to implement in the `BaseEntity` class/etc.

Thoughts?

**Comment by AN-BC** [comment]
Also we have something called `ValueListType` in the EF but don't validate against it at the biz object level other than via strong typing. The `Set` method could bypass that and nothing stops an invalid value, while we're at it in the context of new validation stuff, we should have the `BaseEntity.Validate()` method properly validate all enum values that have ValueListType set to only allow one of the specified values check for this, OR alternatively do this in the `Set()` method, the latter might be better

**Comment by JS-BC** [comment]
@AN-BC Those 4 new fields sounds good to me on first thought. Though what should happen if a given field has a check constraint e.g.  value >= 0 or value <= 10, but in the entity field metadata the min value is set to 1 and max set to 9?

**Comment by AN-BC** [comment]
yeah if check constraints are inconsistent with metadata that won't help. Ideally we parse the CHECK constraint (prob with an LLM) to determine if there is any range based or length based validation in the CHECK statement and if so, derive from that the metadata to interpret at run time. Thoughts?

**Comment by AN-BC** [comment]
@JS-BC  do you think the above solves that concern?

**Comment by JS-BC** [comment]
@AN-BC Yes, the above solves the concern.

**Comment by AN-BC** [comment]
This is closed now because we implemented LLM based parsing of any type of CHECK constraint that ties to validation so a DBA can add a `CHECK` constraint that does any of the things contemplated here as well as anything else that can be represented in any valid boolean expression that can go in there.
